runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -808,6 +834,31 @@ Java_java_lang_invoke_MethodHandleNatives_resolve(JNIEnv *env, jclass clazz, job
 				}
 
 				if (NULL != declaringClass) {
+					if ((NULL != callerClass)
+#if JAVA_SPEC_VERSION >= 16
+					 && (J9_ARE_NO_BITS_SET(lookupMode, MN_UNCONDITIONAL_MODE))
+#endif
+					) {
+						if (callerClass->classLoader != declaringClass->classLoader) {
+							J9BytecodeVerificationData *verifyData = vm->bytecodeVerificationData;
+							omrthread_monitor_enter(vm->classTableMutex);
+							if(0 != verifyData->checkClassLoadingConstraintForNameFunction(
+								currentThread,
+								declaringClass->classLoader,
+								callerClass->classLoader,
+								J9UTF8_DATA(signature),
+								J9UTF8_DATA(signature),
+								J9UTF8_LENGTH(signature),
+								true)
+							) {
+								omrthread_monitor_exit(vm->classTableMutex);
+								vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR, NULL);
+								goto done;
+							}
+							omrthread_monitor_exit(vm->classTableMutex);
```

Comment:
WIth a bit of re-arranging, the lock can be acquired and released at only one place, making it a little easier to reason about the locking.

```suggestion
							omrthread_monitor_enter(vm->classTableMutex);
							J9BytecodeVerificationData *verifyData = vm->bytecodeVerificationData;
							UDATA clConstraintResult = verifyData->checkClassLoadingConstraintForNameFunction(
								currentThread,
								declaringClass->classLoader,
								callerClass->classLoader,
								J9UTF8_DATA(signature),
								J9UTF8_DATA(signature),
								J9UTF8_LENGTH(signature),
								true);
							omrthread_monitor_exit(vm->classTableMutex);
							
							if (0 != clConstraintResult) {
								vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR, NULL);
								goto done;
							}
```